### PR TITLE
Classical Algebra proof-of-concept

### PR DESCRIPTION
**Pull request by Gedol**
Introduces the Monoid and Group types, as well as a short example of Naturals as a monoid.

### CODE REVIEW

File: Lean4Axiomatic/ClassicalAlgebra/Monoid.lean [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+Here we introduce a Monoid using a multiplicative notation.
+
+Note that we only require a binary operation that is associative and
+has identities.  For example, the natural numbers with the addition
+operator is a monoid.
+-/
+
+/-! ### Definitions -/
+
+class Monoid.Ops (α : Type) [EqvOp α] where
```
Comment (cruhland) [inline Lean4Axiomatic/ClassicalAlgebra/Monoid.lean]: The `[EqvOp α]` argument isn't used here, so it can be removed

File: Lean4Axiomatic/ClassicalAlgebra/Monoid.lean [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  f : α → α → α
+  e : α
```
Comment (cruhland) [inline Lean4Axiomatic/ClassicalAlgebra/Monoid.lean]: These names are following mathematical conventions, but they are also very generic. Could you add documentation comments to provide a bit more context on each of them?

Or, maybe we can find alternative names. The `f` operation isn't too bad, because `· * ·` is always used instead. Some thoughts:

```
  f : α → α → α
  ε : α -- epsilon; inspired by the syntax for empty string from CS / formal languages
```

```
  merge : α → α → α
  empty : α
```

```
  binop : α → α → α
  ident : α
```

File: Lean4Axiomatic/ClassicalAlgebra/Monoid.lean [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+export Monoid.Ops (f e)
```
Comment (cruhland) [inline Lean4Axiomatic/ClassicalAlgebra/Monoid.lean]: To encourage use of the `· * ·` operator, let's avoid exporting `f` here. When defining the operator we can use a fully-qualified reference:

```
  mul := Monoid.Ops.f
```

File: Lean4Axiomatic/ClassicalAlgebra/Monoid.lean [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+/-- Enables the use of the `· * ·` operator for f (the monoid operation). -/
+local instance monoid_mul_op_inst {α : Type} [EqvOp α] [Monoid.Ops α] : Mul α := {
```
Comment (cruhland) [inline Lean4Axiomatic/ClassicalAlgebra/Monoid.lean]: The `[EqvOp α]` argument can be dropped here as well

File: Lean4Axiomatic/ClassicalAlgebra/Monoid.lean (hunk 1 above)
Comment (cruhland) [inline Lean4Axiomatic/ClassicalAlgebra/Monoid.lean]: Add a documentation comment for this class

File: Lean4Axiomatic/ClassicalAlgebra/Monoid.lean [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+  mul := f
+}
+
+class Monoid.Props (α : Type) [EqvOp α] [Ops α] where
```
Comment (cruhland) [inline Lean4Axiomatic/ClassicalAlgebra/Monoid.lean]: Add a documentation comment for this class

File: Lean4Axiomatic/ClassicalAlgebra/Monoid.lean (hunk 1 above)
Comment (cruhland) [inline Lean4Axiomatic/ClassicalAlgebra/Monoid.lean]: Replace `where` with `:=`, to stay consistent with other class definitions

File: Lean4Axiomatic/ClassicalAlgebra/Monoid.lean (hunk 5 above)
Comment (cruhland) [inline Lean4Axiomatic/ClassicalAlgebra/Monoid.lean]: Replace `where` with `:=`, to stay consistent with other class definitions

File: Lean4Axiomatic/ClassicalAlgebra/Monoid.lean [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+  substL {x y z : α} : x ≃ y → x * z ≃ y * z
+  substR {x y z : α} : x ≃ y → z * x ≃ z * y
+  assoc {x y z : α} : (x * y) * z ≃ x * (y * z)
+  identityL {x : α} : e * x ≃ x
+  identityR {x : α} : x * e ≃ x
```
Comment (cruhland) [inline Lean4Axiomatic/ClassicalAlgebra/Monoid.lean]: The names `identL`, `identR` would be consistent with other code

File: Lean4Axiomatic/ClassicalAlgebra/Monoid.lean [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+export Monoid.Props (substL substR assoc identityL identityR)
```
Comment (cruhland) [inline Lean4Axiomatic/ClassicalAlgebra/Monoid.lean]: This exports these names into the `CA` namespace, so they could end up clashing with other properties, e.g. the similar ones on `Group`. We probably want to introduce a `CA.Monoid` namespace that they can stay under. But also prevent the `Monoid` class from being put under the `Monoid` namespace, which would make it more awkward to import.

One approach to doing that is to wrap everything in the `Monoid` namespace, except for a small gap where the `Monoid` class is defined:

```
namespace Lean4Axiomatic.CA.Monoid

class Ops ...
class Props ...
export Props (substL ...)

end Monoid

class Monoid ...
attribute [instance] Monoid.toOps
attribute [instance] Monoid.toProps

namespace Monoid

/-! ### Properties -/

variable {α : Type} [EqvOp α] [m : Monoid α]

...

end Lean4Axiomatic.CA.Monoid
```

Another approach is to qualify the names as you've done, but briefly introduce the `Monoid` namespace just for the export:

```
namespace Lean4Axiomatic.CA
...
class Monoid.Props ...
namespace Monoid
export Props (substL ...)
end Monoid

...
end Lean4Axiomatic.CA
```

File: Lean4Axiomatic/ClassicalAlgebra/Monoid.lean [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+class Monoid (α : Type) [EqvOp α] where
```
Comment (cruhland) [inline Lean4Axiomatic/ClassicalAlgebra/Monoid.lean]: Replace `where` with `:=`, to stay consistent with other class definitions

Comment (cruhland) [inline Lean4Axiomatic/ClassicalAlgebra/Monoid.lean]: Add documentation comment to class

File: Lean4Axiomatic/ClassicalAlgebra/Monoid.lean [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+
+attribute [instance] Monoid.toOps
+attribute [instance] Monoid.toProps
+
+/-! ### Properties -/
+
+variable {α : Type} [EqvOp α] [m : Monoid α]
+
+/-- Enables the use of `AA.substL`, `AA.substR`, etc. -/
+instance monoid_subst_inst : AA.Substitutive₂ (α := α) (f) AA.tc (· ≃ ·) (· ≃ ·) := {
```
Comment (cruhland) [inline Lean4Axiomatic/ClassicalAlgebra/Monoid.lean]: You can replace `(f)` with `(· * ·)`, but depending on how you've set up the `Monoid` namespace, you might need to make `local instance monoid_mul_op_inst` into a `scoped instance` instead.

Comment (cruhland) [inline Lean4Axiomatic/ClassicalAlgebra/Monoid.lean]: Please update the line to fit within an 80-char margin:

```suggestion
instance monoid_subst_inst
    : AA.Substitutive₂ (α := α) (f) AA.tc (· ≃ ·) (· ≃ ·)
    := {
```

File: Lean4Axiomatic/ClassicalAlgebra/Monoid.lean [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+  substitutiveL := { subst₂ := λ (_ : True) => substL }
+  substitutiveR := { subst₂ := λ (_ : True) => substR }
+}
+
+/-- There is only one element, namely the identity e, such that e * x ≃ e
+for all elements x. -/
```
Comment (cruhland) [inline Lean4Axiomatic/ClassicalAlgebra/Monoid.lean]: If a comment wraps over multiple lines, I like to put the start and end delimiters by themselves:

```suggestion
/--
There is only one element, namely the identity e, such that e * x ≃ e
for all elements x.
-/
```

File: Lean4Axiomatic/ClassicalAlgebra/Monoid.lean [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+def mul_identity_unique
```
Comment (cruhland) [inline Lean4Axiomatic/ClassicalAlgebra/Monoid.lean]: It would be more appropriate to declare this a `theorem` instead of a `def`, because it proves a property (the return type is in `Prop`) instead of returning data (the return type is not in `Type`).

File: Lean4Axiomatic/ClassicalAlgebra/Monoid.lean [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+    {x : α} (x_is_left_identity : ((y : α) → (x * y) ≃ y)) :
+    x ≃ e :=
+  calc
```
Comment (cruhland) [inline Lean4Axiomatic/ClassicalAlgebra/Monoid.lean]: My rule of thumb is to put the signature on as few lines as possible, while making minor adjustments for aesthetics. If a tactic is used to introduce the proof, that can be put on the same line as `:=`:

```suggestion
    {x : α} (x_is_left_identity : ((y : α) → (x * y) ≃ y)) : x ≃ e
    := calc
```

File: Lean4Axiomatic/ClassicalAlgebra/Monoid.lean [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+/-- Enables the use of `AA.substL`, `AA.substR`, etc. -/
+instance monoid_subst_inst : AA.Substitutive₂ (α := α) (f) AA.tc (· ≃ ·) (· ≃ ·) := {
+  substitutiveL := { subst₂ := λ (_ : True) => substL }
+  substitutiveR := { subst₂ := λ (_ : True) => substR }
+}
+
+/-- There is only one element, namely the identity e, such that e * x ≃ e
+for all elements x. -/
+def mul_identity_unique
+    {x : α} (x_is_left_identity : ((y : α) → (x * y) ≃ y)) :
```
Comment (cruhland) [inline Lean4Axiomatic/ClassicalAlgebra/Monoid.lean]: The type of `x_is_left_identity` doesn't seem to match the documentation comment

File: Lean4Axiomatic/ClassicalAlgebra/Monoid.lean [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+    x ≃ e :=
+  calc
+    x    ≃   _ := Rel.symm identityR
+    x * e ≃  _ := x_is_left_identity e
+    e ≃      _ := Rel.refl
```
Comment (cruhland) [inline Lean4Axiomatic/ClassicalAlgebra/Monoid.lean]: Current convention is to put the `_ ≃` portion on the left. The first line can use `_ =` and `:= rfl`. Align everything:

```suggestion
  _ = x     := rfl
  _ ≃ x * ε := Rel.symm identR
  _ ≃ ε     := x_is_left_identity ε
```
(this uses `ε` instead of `e`)

**Review by cruhland — COMMENTED** [review]
Looks good! Added some comments to `Monoid`, take a look and let me know what you think. If you agree with the feedback, see if any of it can be applied to `Group` as well.

File: Lean4Axiomatic/ClassicalAlgebra/Monoid.lean (hunk 7 above)
Comment (Gedol) [reply Lean4Axiomatic/ClassicalAlgebra/Monoid.lean]: I had some trouble implementing your suggestion.  What I ended up doing is something similar to your first approach and put everything in namespace Lean4Axiomatic.CA.  I renamed Monoid.{Props,Ops} to {Props,Ops}.  I kept Monoid class.

This didn't have any issues and it seems ok to import Monoid as  CA.Monoid.Monoid.

File: Lean4Axiomatic/ClassicalAlgebra/Monoid.lean (hunk 13 above)
Comment (Gedol) [reply Lean4Axiomatic/ClassicalAlgebra/Monoid.lean]: At first I wasn't sure of the problem as the definition and comment  matches my intuition of what I'm proving.

One minor thing is I used term $x$ in two places with different meanings, which could be confusing and that's easy to fix.

I then realized there is a subtler issue, which is I didn't prove that $e$ is unique b/c " ≃ " is an EqvOp.  So I added a comment to address.  Let me know if you think it looks alright.  (or if I missed your point completely let me know).

File: Lean4Axiomatic/ClassicalAlgebra/Monoid.lean [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+  binop : α → α → α
+  ident : α
+export Ops (ident)
+
+/-- Enables the use of the `· * ·` operator for f (the monoid operation). -/
+local instance monoid_mul_op_inst {α : Type} [Monoid.Ops α] : Mul α := {
+  mul := Monoid.Ops.binop
+}
+
+/-- Properties of rational number multiplication. -/
```
Comment (cruhland) [inline Lean4Axiomatic/ClassicalAlgebra/Monoid.lean]: Looks like a copy-paste?

File: Lean4Axiomatic/ClassicalAlgebra/Monoid.lean [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+class Props (α : Type) [EqvOp α] [Ops α] :=
+  substL {x y z : α} : x ≃ y → x * z ≃ y * z
+  substR {x y z : α} : x ≃ y → z * x ≃ z * y
+  assoc {x y z : α} : (x * y) * z ≃ x * (y * z)
+  identL {x : α} : ident* x ≃ x
```
Comment (cruhland) [inline Lean4Axiomatic/ClassicalAlgebra/Monoid.lean]: Spacing

```suggestion
  identL {x : α} : ident * x ≃ x
```

File: Lean4Axiomatic/ClassicalAlgebra/Monoid.lean (hunk 13 above)
Comment (cruhland) [reply Lean4Axiomatic/ClassicalAlgebra/Monoid.lean]: > I then realized there is a subtler issue, which is I didn't prove that $e$ is unique b/c " ≃ " is an EqvOp.  So I added a comment to address.  Let me know if you think it looks alright.

Nah, that part was fine, I assume that any reader of the code understands that our equivalence operation `≃` is the default notion of identity. So uniqueness is always meant in terms of `≃` unless stated otherwise. It's okay to leave out the "Technical note".

Comment (cruhland) [reply Lean4Axiomatic/ClassicalAlgebra/Monoid.lean]: What I meant with my original message was that your documentation comment says:
* the element `x` satisfies the property `x * y ≃ x` for all elements `y`

while the `x_is_left_ident` argument to the theorem says:
* the element `x` satisfies the property `x * y ≃ y` for all elements `y`

and these are not the same.

File: Lean4Axiomatic/ClassicalAlgebra/Group.lean [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+/-! ### Definitions -/
+
+/--
+Operations for Group, namely the binary operation, identity element, and
+existence of inverses.
+-/
+class Ops (α : Type) :=
+  binop : α → α → α
+  ident : α
+  inverse : (x : α) → α
```
Comment (cruhland) [inline Lean4Axiomatic/ClassicalAlgebra/Group.lean:22]: Using `inverse` directly can get a bit wordy, so it might be a good idea to introduce notation for it. You can probably just copy what I do in `Rational/Reciprocation`:

```
/-- Enables the use of the `·⁻¹` operator for reciprocation. -/
postfix:120 "⁻¹" => reciprocal
```

File: Lean4Axiomatic/ClassicalAlgebra/Group.lean [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+
+export Ops (ident inverse)
+
+/-- Enables the use of the `· * ·` operator for f (the monoid operation). -/
```
Comment (cruhland) [inline Lean4Axiomatic/ClassicalAlgebra/Group.lean]: This comment needs to be updated for groups:

```suggestion
/-- Enables the use of the `· * ·` operator for binop. -/
```

File: Lean4Axiomatic/ClassicalAlgebra/Group.lean [hunk 19]
```diff
@@ -0,0 +1,108 @@
...
+
+/-- Properties of Group. -/
+class Props (α : Type) [EqvOp α] [Ops α] :=
+  substL {x y z : α} : x ≃ y → x * z ≃ y * z
+  substR {x y z : α} : x ≃ y → z * x ≃ z * y
+  assoc {x y z : α} : (x * y) * z ≃ x * (y * z)
+  identL {x : α} : ident * x ≃ x
+  identR {x : α} : x * ident ≃ x
+  inverse_propL (x : α) : (inverse x) * x ≃ ident
+  inverse_propR (x : α) : x * (inverse x) ≃ ident
```
Comment (cruhland) [inline Lean4Axiomatic/ClassicalAlgebra/Group.lean]: I think `inverseL` and `inverseR` are good names for these.

File: Lean4Axiomatic/ClassicalAlgebra/Group.lean [hunk 20]
```diff
@@ -0,0 +1,108 @@
...
+  substitutiveR := { subst₂ := λ (_ : True) => substR }
+}
+
+/--
+You May perform cancellation of an element x, and conclude from
+x * y ≃ x * z that y ≃ z.
+-/
+def group_cancellationL
+    {x y z : α} : (x * y ≃ x * z) → y ≃ z := λ _ =>
+  show y ≃ z from calc
```
Comment (cruhland) [inline Lean4Axiomatic/ClassicalAlgebra/Group.lean]: This should be a `theorem`. Naming it `group_cancelL` is consistent with other code. The parentheses around the hypothesis are redundant. Also, instead of lambda notation, I use the `intro` tactic:

```suggestion
theorem group_cancelL {x y z : α} : x * y ≃ x * z → y ≃ z := by
  intro (_ : x * y ≃ x * z)
  show y ≃ z
  calc
```
Be sure to add an extra indent to the lines after the `calc` tactic!

File: Lean4Axiomatic/ClassicalAlgebra/Group.lean [hunk 21]
```diff
@@ -0,0 +1,108 @@
...
+    _ ≃ (inverse x) * (x * y) := assoc
+    _ ≃ (inverse x) * (x * z) := substR ‹x * y ≃ x * z›
+    _ ≃ (inverse x * x) * z   := Rel.symm assoc
+    _ ≃ ident * z             := substL (inverse_propL x)
+    _ ≃ z                     := identL
+
+
+local instance monoid_from_group_ops :  CA.Monoid.Ops α := {
+  binop := (· * ·)
+  ident := g.toOps.ident
```
Comment (cruhland) [inline Lean4Axiomatic/ClassicalAlgebra/Group.lean]: I think you should be able to use plain `ident` here, or maybe `Ops.ident`?

File: Lean4Axiomatic/ClassicalAlgebra/Group.lean [hunk 22]
```diff
@@ -0,0 +1,108 @@
...
+Demonstrates that any group is also a monoid.
+-/
+instance monoid_from_group : CA.Monoid.Monoid (α := α) := {
+  toOps := monoid_from_group_ops
+  toProps := {
+    substL    := g.toProps.substL
+    substR    := g.toProps.substR
+    assoc     := g.toProps.assoc
+    identL := g.toProps.identL
+    identR := g.toProps.identR
```
Comment (cruhland) [inline Lean4Axiomatic/ClassicalAlgebra/Group.lean]: Similarly, in these lines I think you can use e.g. `substL`, or at most `Props.substL`, etc.

File: Lean4Axiomatic/ClassicalAlgebra/Group.lean [hunk 23]
```diff
@@ -0,0 +1,108 @@
...
+}
+
+/--
+  There is only one element, namely the identity ident, such that
+  ident * y ≃ ident for all elements y.  The proof follows from the fact that
+  Groups are Monoids and the result holds for Monoids.
+-/
+theorem mul_identity_unique
+    {x : α} (x_is_left_ident : ((y : α) → (x * y) ≃ y)) : x ≃ ident :=
+  Lean4Axiomatic.CA.Monoid.mul_identity_unique x_is_left_ident
```
Comment (cruhland) [inline Lean4Axiomatic/ClassicalAlgebra/Group.lean]: This is a good demonstration that the `mul_identity_unique` theorem for monoids can be extended to groups, but it won't really be useful on its own, because we can always just use the monoid version of it. Maybe you can make it an unnamed `example` instead of a `theorem`?

File: Lean4Axiomatic/ClassicalAlgebra/Monoid.lean [hunk 24]
```diff
@@ -0,0 +1,84 @@
...
+-/
+class Ops (α : Type) :=
+  binop : α → α → α
+  ident : α
+export Ops (ident)
+
+/-- Enables the use of the `· * ·` operator for f (the monoid operation). -/
+local instance monoid_mul_op_inst {α : Type} [Monoid.Ops α] : Mul α := {
+  mul := Monoid.Ops.binop
+}
```
Comment (cruhland) [inline Lean4Axiomatic/ClassicalAlgebra/Monoid.lean]: Nitpick - the `Monoid` prefix is redundant here:

```suggestion
local instance monoid_mul_op_inst {α : Type} [Ops α] : Mul α := {
  mul := Ops.binop
}
```

File: Lean4Axiomatic/ClassicalAlgebra/Group.lean [hunk 25]
```diff
@@ -0,0 +1,108 @@
...
+
+attribute [instance] Group.toOps
+attribute [instance] Group.toProps
+
+/-! ### Properties -/
+
+variable {α : Type} [EqvOp α] [g : Group α]
+
+/-- Enables the use of `AA.substL`, `AA.substR`, etc. -/
+instance group_subst_inst
```
Comment (cruhland) [inline Lean4Axiomatic/ClassicalAlgebra/Group.lean]: You should make this a `local instance`, to avoid introducing a conflicting instance value when defining a group on a type that already has a `Substitutive₂` instance for the binary operator (such as the natural numbers).

File: Lean4Axiomatic/ClassicalAlgebra/Monoid.lean [hunk 26]
```diff
@@ -0,0 +1,84 @@
...
+attribute [instance] Monoid.toOps
+attribute [instance] Monoid.toProps
+
+
+/-! ### Properties -/
+
+variable {α : Type} [EqvOp α] [m : Monoid α]
+
+/-- Enables the use of `AA.substL`, `AA.substR`, etc. -/
+instance monoid_subst_inst
```
Comment (cruhland) [inline Lean4Axiomatic/ClassicalAlgebra/Monoid.lean]: This should be a `local instance` too

File: Lean4Axiomatic/Natural/Addition.lean [hunk 27]
```diff
@@ -292,4 +293,34 @@ theorem zero_sum_split {n m : ℕ} : n + m ≃ 0 → n ≃ 0 ∧ m ≃ 0 := by
...
+local instance add_monoid_ops :  CA.Monoid.Ops ℕ := {
+  binop := (· + ·)
+  ident := 0
+}
+
+def subst_addR {n₁ n₂ m : ℕ} : n₁ ≃ n₂ → m + n₁ ≃ m + n₂ := AA.substR
+
+def add_monoid_props : CA.Monoid.Props ℕ := {
+  substL   := subst_add
+  substR   := subst_addR
```
Comment (cruhland) [inline Lean4Axiomatic/Natural/Addition.lean]: Since `Substitutive₂` is already defined for naturals, can you just use `AA.substL` and `AA.substR` here? It might require you to first change your group and monoid instances of `Substitutive₂` into _local_ instances.

```suggestion
def subst_addR {n₁ n₂ m : ℕ} : n₁ ≃ n₂ → m + n₁ ≃ m + n₂ := AA.substR

def add_monoid_props : CA.Monoid.Props ℕ := {
  substL   := AA.substL
  substR   := AA.substR
```

**Review by cruhland — COMMENTED** [review]
Looking good! Just a few more comments. Also, can you look through older comments and make sure none were missed?

**Comment by Gedol** [comment]
Thanks Charles,

I am running into issues with you last suggest change in addition:

def add_monoid_props : CA.Monoid.Props ℕ := {
  substL   := AA.substL

The 2nd line above gives error

failed to synthesize
AA.SubstitutiveOn Hand.L HMul.hMul AA.tc Operators.tildeDash
Operators.tildeDash
use `set_option diagnostics true` to get diagnostic informationLean 4


substL is of type
substL : ∀ {x y z : ℕ}, x ≃ y → x * z ≃ y * z


So maybe Lean is having difficulty telling that the instance of

instance add_substitutive
    : AA.Substitutive₂

provides evidence of this.

how would you proceed here?

thanks,
Alex

**Comment by Gedol** [comment]
the following is the closest I could get to just using AA.substL:

def add_monoid_props : CA.Monoid.Props ℕ :=
  let subst_addL {n₁ n₂ m : ℕ} : n₁ ≃ n₂ → n₁ + m ≃ n₂ + m := AA.substL;
  let subst_addR {n₁ n₂ m : ℕ} : n₁ ≃ n₂ → m + n₁ ≃ m + n₂ := AA.substR;
{
  substL  := subst_addL
  substR   := subst_addR
  assoc    := add_associative.assoc
  identL := add_identity.identityL.ident
  identR := add_identity.identityR.ident
}


if I try to use AA.substL directly for the substL field above lean gives
the "failed to synthesis" error I pasted above.  It seems like Lean has
trouble figuring out that the relevant monoid operation is '+' if we just
use AA.substL directly, but it's fine when given subst_addL.  If you hover
above substL above it gives the type:


substL : ∀ {x y z : ℕ}, x ≃ y → x + z ≃ y + z


which has replaced the '*' with '+' as we intend.

However if I replace the line above with
  substL := sorry
or
  substL := AA.substL

it lists the type as

substL : ∀ {x y z : ℕ}, x ≃ y → x * z ≃ y * z


asdf
So I don't really understand the behavior here.

Alex
